src/utils/AssayGeneratorMetabo.cpp

```diff
@@ -0,0 +1,523 @@
+
+        // calculation of contrast angle (cosine simiarity)
+        for (std::vector<size_t>::iterator index_it = index.begin(); index_it != index.end(); ++index_it)
+        {
+          const MSSpectrum &spectrum = spectra[*index_it];
+          const BinnedSpectrum binned_spectrum(spectrum, BinnedSpectrum::DEFAULT_BIN_WIDTH_HIRES, false, 1);
+
+          BinnedSpectralContrastAngle bspa;
+          double cosine_sim = bspa(binned_highest_int, binned_spectrum);
+
+          if (cosine_sim > cosine_sim_threshold)
+          {
+            similar_spectra.push_back(spectrum);
+            exp.addSpectrum(spectrum);
+          }
+        }
+        // calculate consensus spectrum
+        exp.sortSpectra();
+        SpectraMerger merger;
+        Param p;
+        p.setValue("precursor_method:mz_tolerance", precursor_mz_tol*2);
+        p.setValue("precursor_method:rt_tolerance", precursor_rt_tol*2);
+        merger.setParameters(p);
+
+        // all MS spectra should have the same precursor
+        merger.mergeSpectraPrecursors(exp);
+
+        // check if all precursors have been merged if not use highest intensity precursor
+        if (exp.getSpectra().size() < 2)
+        {
+          transition_spectrum = exp.getSpectra()[0];;
+        }
+      }
+
+      // transition calculations
+      // calculate max intensity peak and threshold
+      float max_int = 0.0;
+      float min_int = numeric_limits<float>::max();
+      for (MSSpectrum::const_iterator spec_it = transition_spectrum.begin(); spec_it != transition_spectrum.end(); ++spec_it)
+      {
+        //find the max intensity peak
+        if (spec_it->getIntensity() > max_int)
+        {
+          max_int = spec_it->getIntensity();
+        }
+        if (spec_it->getIntensity() < min_int)
+        {
+          min_int = spec_it->getIntensity();
+        }
+      }
+
+      // no peaks or all peaks have same intensity (single peak / noise)
+      if (min_int >= max_int) { continue; }
+
+      // threshold should be at x % of the maximum intensity
+      // hard minimal threshold of min_int * 1.1
+      float threshold_transition = max_int * (transition_threshold/100);
+      float threshold_noise = min_int * 1.1;
+
+      AssayRow row;
+      int transition_counter = 1;
+
+      for (MSSpectrum::iterator spec_it = transition_spectrum.begin(); spec_it != transition_spectrum.end(); ++spec_it)
+      {
+        float current_int = spec_it->getIntensity();
+        double current_mz = spec_it->getMZ();
+
+        // write row for each transistion
+        // current int has to be higher than transition thresold and should not be smaller than threshold noise
+        if (current_int > threshold_transition && current_int > threshold_noise)
+        {
+          float rel_int = current_int/max_int;
+          row.precursor_mz = highest_precursor_mz;
+          row.product_mz = current_mz;
+          row.library_int = rel_int;
+          row.normalized_rt = spectrum_rt;
```

Comment:
I think it may be smarter to take the *Feature* RT here since this is simply the RT of where the MS2 spectrum was triggered. Due to dynamic exclusion, this may be far from the apex of the actual RT peak.